Feature/update batteries

oscarjaergren opened the pull request.

KarlKroll committed:
Update repo to dotnet 9

KarlKroll committed:
Update the gitversion

KarlKroll committed:
Downgrade X nit to when it worked

KarlKroll committed:
Temp commit

KarlKroll committed:
Just to try it

KarlKroll committed:
Test

KarlKroll committed:
Log

KarlKroll committed:
Try updating cake

KarlKroll committed:
Try

KarlKroll committed:
Update cake

KarlKroll committed:
Might as well try

KarlKroll committed:
Try

KarlKroll committed:
Try

KarlKroll committed:
Try

KarlKroll committed:
Try

KarlKroll committed:
Fix eof

KarlKroll committed:
Add debuggign

KarlKroll committed:
Try

KarlKroll committed:
Try

KarlKroll committed:
Try

KarlKroll committed:
Try publish

KarlKroll committed:
Try

KarlKroll committed:
Force include

KarlKroll committed:
Try

KarlKroll committed:
Dumb attempt

KarlKroll committed:
Msbuild

KarlKroll committed:
try

oscarjaergren committed:
Remove cake and use github actions

oscarjaergren committed:
Update CI

oscarjaergren committed:
Add permissions

oscarjaergren committed:
Add credentials

oscarjaergren committed:
Try this

oscarjaergren committed:
Add GitHub Actions workflow for publishing .NET packages

KarlKroll committed:
Allow manual triggering

KarlKroll committed:
Test

KarlKroll committed:
Try

KarlKroll committed:
Try

KarlKroll committed:
Test

KarlKroll committed:
Try

KarlKroll committed:
Try it

KarlKroll committed:
Try

KarlKroll committed:
Refactor CI workflows and update GitVersion tags for consistency

KarlKroll committed:
Add support for pull request branches in GitVersion configuration

oscarjaergren committed:
Update CI workflow to conditionally pack NuGet packages with version suffix for non-main branches

oscarjaergren committed:
Set fail-fast to false in CI workflow to improve job execution reliability

KarlKroll committed:
Just use publish and just Linux for now

KarlKroll committed:
Merge branch 'feature/update-batteries' of https://github.com/crispthinking/IronRe2 into feature/update-batteries

KarlKroll committed:
Fix conflict

KarlKroll committed:
Comment out ci temporarily

KarlKroll committed:
Fix pack for publish

KarlKroll committed:
Update package versioning in publish workflow

KarlKroll committed:
Update package version in publish workflow and adjust IronRe2-Batteries.Linux reference

KarlKroll committed:
Refactor regex compilation to handle options more cleanly

KarlKroll committed:
Enhance UnmanagedResource class with improved disposal pattern and null handling for raw handle

KarlKroll committed:
Add error handling for regex compilation to throw exception on unknown errors

KarlKroll committed:
Update IronRe2-Batteries.Linux package version to 0.1.1-PullRequest6.131

KarlKroll committed:
Add IronRe2-Batteries.OSX package reference version 0.1.1-PullRequest6.131

KarlKroll committed:
Add macOS support to GitHub Actions workflow

oscarjaergren committed:
Debug code for Release Handle

oscarjaergren committed:
Lint

oscarjaergren committed:
More clean up

KarlKroll committed:
Add DotMemoryUnit attribute to MemoryLeakTests

KarlKroll committed:
Switch from DLL Import to LibraryImport

KarlKroll committed:
Merge remote-tracking branch 'origin/feature/update-batteries' into feature/update-batteries

abbottdev commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,71 @@
+
+      - name: Pack
+        if: ${{ matrix.os == 'ubuntu-latest' }}
+        run: dotnet pack IronRe2.sln --configuration Release --no-build --output Artifacts
```
I think we should be publishing symbols for each os type right? I'm not sure the matrix os is enough here to switch the platforms we build packages for.

We probably want something like:
```
dotnet pack --runtime win-x64
dotnet pack --runtime linux-x64
dotnet pack --runtime linux-arm64
dotnet pack --runtime osx-arm64
```

I _think_ you should be able to safely do that in a single step outside of a matrix, because the `rid` flag affects what packages are used for restoring.  (We may want it to be part of a matrix variable)

iwillspeak replied:
We're not packing for a specific runtime here James. This is the managed code. It is just predicated like this to make sure that only one part of the matrix tries to publish.

oscarjaergren replied:
I considered this, but that will result in there being 4 different nuget packages won't it? Which is fine I guess

iwillspeak replied:
It isn't fine @oscarjaergren . We don't want any platform specific code in the main Re2 package. Building once is correct here.

oscarjaergren replied:
@iwillspeak What should be done with this exactly?

oscarjaergren replied:
Leaving as is then.

abbottdev commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,45 @@
+#       - published # Run the workflow when a new GitHub release is published
+#   workflow_dispatch: # Allow manual triggering from the GitHub UI
+
+    # TEMP MAKE IT RUN SO I CAN TEST PUBLISH
```
still temp? 🤓

oscarjaergren replied:
Removed

abbottdev commented on nuget.config:
```diff
@@ -2,9 +2,6 @@
   <packageSources>
     <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
-    <add key="crisp.prerelease" value="https://crisp-thinking.pkgs.visualstudio.com/_packaging/PreRelease/nuget/v3/index.json" />
+<!--    <add key="github" value="https://nuget.pkg.github.com/crispthinking/index.json" />-->
```
should we just drop this?

iwillspeak replied:
the pre-release feed was a disabled package soruce before. Useful to have so it can easily be enabled I guess? But you're right, we want the default source we build against to be the publich nuget one.

oscarjaergren replied:
Adjusted

abbottdev commented on src/Directory.Build.props:
```diff
@@ -3,12 +3,14 @@
         <nullable>enable</nullable>
         <GenerateDocumentationFile>true</GenerateDocumentationFile>
+
+        <!-- TEMPORARILY DISABLE -->
```
todo?

oscarjaergren replied:
Removed

abbottdev commented on src/IronRe2/ByteRange.cs:
```diff
@@ -1,17 +1,16 @@
+///  code and store the result in this structure instead.
+/// </para>
+/// </summary>
+internal struct ByteRange
```
I think we should be able to substitute this with `System.Range` now it's a "thing"

iwillspeak replied:
I'd rather keep it separate, but perhaps we could add implicit ops to and from Range?

oscarjaergren replied:
Attempted to enchance it

abbottdev commented on src/IronRe2/Re2Handle.cs:
```diff
@@ -1,22 +1,65 @@
+        : base(IntPtr.Zero, true)
+    {
+        // Log creation of a new instance for debugging purposes.
+        Debug.WriteLine($"[Re2Handle] Created instance with handle: {handle}");
```
Needs compiler conditional `#if DEBUG` or just drop it.

oscarjaergren replied:
Removed debug code

abbottdev commented on src/IronRe2/Re2Handle.cs:
```diff
@@ -1,22 +1,65 @@
+        // Log that we're about to release a handle.
+        if (!IsInvalid)
         {
+            Debug.WriteLine($"[Re2Handle] Releasing handle: {handle}");
```
Can we wrap all of this debug code up into a compiler constant `#if DEBUG` or similar?

oscarjaergren replied:
Best removing it

oscarjaergren replied:
Removed debug code

abbottdev commented on src/IronRe2/RegexSet.cs:
```diff
@@ -3,154 +3,150 @@
+        IReadOnlyCollection<byte[]> patternsAsBytes,
+        Options options)
+    {
+        // TODO: we could maybe have a `RegexSetBuilder` to represent this
```
TODO

iwillspeak replied:
Old TODO, it has just been moved. 

As for builders Crisp has one, but I'm happy leaving this API as it is, with a set being constructed with a list of patterns and options.

oscarjaergren replied:
Not sure what to do with this one, it was like this before and not sure what intent is.

abbottdev reviewed (commented):
A few tidy up's still remain in here 😄

iwillspeak commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,45 @@
+    uses: ./.github/workflows/ci.yml
+
+  publish:
+   #TEMP DISABLE if: github.event_name == 'release'
```
Temp

oscarjaergren replied:
Removed

iwillspeak commented on src/IronRe2/IronRe2.csproj:
```diff
@@ -1,14 +1,17 @@
+		<LangVersion>preview</LangVersion>
+	</PropertyGroup>
+	<ItemGroup>
+		<PackageReference Include="IronRe2-Batteries.Linux" Version="0.1.1-PullRequest6.144" />
```
Needs stable builds, and those builds should probably be properly versioned too.

oscarjaergren replied:
Updated

iwillspeak commented on src/IronRe2/IronRe2.csproj:
```diff
@@ -1,14 +1,17 @@
+		<PackageReference Include="IronRe2-Batteries.Windows" Version="0.1.1-PullRequest6.144" />
+	</ItemGroup>
+	<ItemGroup>
+		<Content Include="runtimes\**\*.*">
```
What runtimes content are we adding into this library?

oscarjaergren replied:
Remove includes

iwillspeak commented on src/IronRe2/NamedCaptureEnumerator.cs:
```diff
@@ -3,60 +3,55 @@
-                return false;
-            }
+            string? name = Marshal.PtrToStringAnsi(new IntPtr(namePtr));
+            if (name != null) _current = new NamedCaptureGroup(name, index);
```
Why has this changed? Don't we want to always set the capture group? Otherwise we'll leave `_current` as the last value right?

iwillspeak commented on src/IronRe2/Re2Handle.cs:
```diff
@@ -1,22 +1,65 @@
+    /// <summary>
+    /// Derived classes implement this to release the unmanaged resource.
+    /// </summary>
+    protected abstract bool ReleaseHandleCore();
```
Can we drop this custom release handle stuff and go back to the old stuff? I'm not sure we want any of the debugging stuff.

iwillspeak commented on src/IronRe2/RegexSetBuilder.cs:
```diff
@@ -1,58 +1,56 @@
-    {
-        private List<byte[]> _patterns = new ();
-        private Options _options;
+    public RegexSetBuilder()
```
Ah, @abbottdev , it looks like that TODO _is_ TODONE.

iwillspeak commented on test/IronRe2.Tests/IronRe2.Tests.csproj:
```diff
@@ -1,19 +1,28 @@
+    <Content Include="bin\Release\net9.0\IronRe2.dll" />
+    <Content Include="bin\Release\net9.0\runtimes\linux-x64\native\libcre2.so" />
+    <Content Include="bin\Release\net9.0\runtimes\osx\native\libcre2.dylib" />
+    <Content Include="bin\Release\net9.0\runtimes\win-x64\native\cre2.dll" />
```
this all needs cleaning up.

oscarjaergren replied:
All done

iwillspeak commented on test/IronRe2.Tests/MetaTests.cs:
```diff
@@ -1,34 +1,32 @@
-            //When
-            var version = Meta.VersionString;
+        //When
+        string version = Meta.VersionString;
```
Curious as to why we've moved away from `var` all over the place.

oscarjaergren replied:
it's personal preference, mainly cause I really like the new() pattern, happy to change back though, way shorter.

oscarjaergren replied:
Reverted

iwillspeak commented on test/IronRe2.Tests/MetaTests.cs:
```diff
@@ -1,34 +1,32 @@
-            //When
-            var version = Meta.Version;
+        //When
+        (int current, int revision, int age) version = Meta.Version;
```
See this move away from `var` just makes the whole thing more complex than it needs to be.

oscarjaergren replied:
it's personal preference, mainly cause I really like the new() pattern, happy to change back though

oscarjaergren replied:
All done

oscarjaergren committed:
Update nuget.config with additional package sources

Added `github` and `crisp.prerelease` package sources.
Introduced `disabledPackageSources` to disable the
`crisp.prerelease` source.

oscarjaergren committed:
Update Directory.Build.props for stricter warnings handling

Removed comment indicating temporary disablement of the
'TreatWarningsAsErrors' property, ensuring that warnings
are treated as errors during the build process for improved
code quality checks.

oscarjaergren committed:
Refactor ByteRange struct for improved usability

The `ByteRange` struct has been updated to be an `internal readonly struct`, enhancing immutability. Properties `Start` and `Past` are now read-only with public getters, and a new `Length` property has been added. A constructor with validation ensures the `past` value is not less than `start`. Implicit conversion operators for `ByteRange` and `System.Range` have been introduced, along with exception handling for unsupported cases. Equality members and a `ToString` method have also been implemented to improve the struct's functionality.

oscarjaergren committed:
Refactor IronRe2.csproj formatting

Reformatted the `<PropertyGroup>` and `<ItemGroup>` sections in `IronRe2.csproj` to remove unnecessary indentation and line breaks. The content remains unchanged, preserving the target framework and package references.

oscarjaergren committed:
Update IronRe2-Batteries package versions

Upgraded `IronRe2-Batteries` packages for Linux, OSX, and Windows in `IronRe2.csproj` from version `0.1.1-PullRequest6.144` to `3.0.0` to incorporate new features, improvements, and bug fixes.

oscarjaergren committed:
Use var everywhere

oscarjaergren committed:
Remove temp code

oscarjaergren committed:
run linter

oscarjaergren committed:
Remove debug code

oscarjaergren committed:
Cleanup

oscarjaergren committed:
Remove add github source, shouldn't be needed

oscarjaergren committed:
See if this makes a difference?

oscarjaergren committed:
Revert

oscarjaergren committed:
Fix ubuntu version so this all works in  later

oscarjaergren committed:
Temp try to see if it fixes it

oscarjaergren committed:
Revert that attempt to fix macOS

oscarjaergren committed:
Update to latest version

KarlKroll committed:
Remove crisp feed and disable github feed

KarlKroll committed:
Clean up

abbottdev commented on src/IronRe2/Regex.cs:
```diff
@@ -1,486 +1,507 @@
+        ReadOnlySpan<byte> pattern, ReadOnlyMemory<byte> haystack)
+    {
+        // Use an explicit one-element array for captures.
+        var captures = new Re2Ffi.cre2_string_t[1];
```
Worth verifying whether captures[0] being null here would be an issue (before it was initialized so it wasn't null, but was still an empty object so _probably_ isn't used.)

iwillspeak replied:
It’s a struct so shouldn’t matter.

abbottdev reviewed (commented):
Just check out that array initialiser 👍
